build pdf in quarto and bookdown - migrate to quarto

mitzimorris opened the pull request:
#### Submission Checklist

- [x] Builds locally 
- [x] New functions marked with `` `r since("VERSION")` ``
- [x] Declare copyright holder and open-source license: see below

#### Summary

Use quarto to build individual pdfs of the 4 docsets  - Stan User's Guide, Reference Manual, Functions Reference and the CmdStan User's Guide; also provide quarto-generated website for all manuals with top nav bar that lets users access all Stan Docs plus interface docs.

Changes:

*  latex tables instead of kablextra if table has mathjax
*  don't use macro '\[` for equation arrays.
*  common `header.tex` and `postamble.tex` (for docs which have indexes) in `src` dir.
*  Table of Contents should have lowercase roman page numbers; main section has arabic numbers starting from 1
*  Ensure proper headers/footers everywhere.  due to difference between quarto and bookdown, add latex directives '\mainmatter' to `index.Rmd` and `index.qmd` and then change pagestyle to `plain` for references.
* light edits to overview.
* ensure that subsections and subsubsections are either consistently numbered (functions-reference, cmdstan-guide) or unnumbered (stan-users-guide, reference-manual).
* Functions Reference - remove hand-generated sets of links for chapter subsections, needed for bookdown html version, which renders each subsection as its own page.

Additions:

* script to build pdf version of the docs using quarto


#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Columbia University



By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC BY-ND 4.0 (https://creativecommons.org/licenses/by-nd/4.0/)

mitzimorris commented:
this PR is only for building individual pdfs with quarto instead of bookdown.

quarto makes a distinction between projects for websites and books.  the per-document `_quarto.yml` files in the `src` directory subdirs are for projects of type `book`.    

to build a quarto-based website which lets users switch between documents via a top navbar, will require a different `_quarto.yml` file, type `website` which will live in the `_src` directory.

mitzimorris commented:
update: 

- to support both pdf versions plus a quarto-based unified docs website, added script `build_quarto_pdfs.py` which handles the quarto build logic.
- added files and logic to `src` dir which will build a quarto website for all the docs.

this is what the landing page looks like:

![docs_landing](https://github.com/stan-dev/docs/assets/5466612/4a5ce7ba-d25d-4100-a9c0-e91bb5889691)

this is what a page of the docs looks like:
![unified_docs](https://github.com/stan-dev/docs/assets/5466612/95516ecb-0979-49a1-9d27-2efee94bc9e8)


now that all docs are in same place, cross references seem to be broken everywhere.   todo.

mitzimorris commented:
update:  this PR now builds both pdfs for individual docs and website for all docs.
ready for review and/or beta testing.

latest change is to remove the html-version hack in the functions reference which provides a per-chapter psuedo-toc in the form of a set of links to the rendered html subsections.   this isn't needed, given quarto right-hand-column toc.   also upped the website toc depth to 3 - this only affects the right-hand-column toc - the sidebar nav shows chapters only.

still need to fix cross-refs - it's unclear whether book and websites both allow the kinds of hyperlinks described here:  https://quarto.org/docs/books/book-crossrefs.html#hyperlinks

the SBC chapter could use some Goldilocks love w/r/t the figures - some render too big, some too small.

mitzimorris commented:
the remaining problem to be solved for migration to quarto is how to update "link to latest" logic and to make sure that the generated html can be served with github pages.

WardBrian commented on build_quarto_pdfs.py:
```diff
@@ -0,0 +1,97 @@
+        shexec(command)
+        outpath = os.path.join(srcpath, '_pdf', pdfname)
+        shutil.move(outpath, docspath)
+        safe_rm(".gitignore")  # quarto cleanup
```
Rather than delete this I think we should just check-in .gitignores in each of these folders. We can also ignore `_pdf/` and `_website/` in the top-level .gitignore

WardBrian commented on build_quarto_pdfs.py:
I'm not sure if a build script is as necessary as it once was, considering how simplified the quarto command line is compared to invoking bookdown from Rscript. 

E.g., we could just describe the build in the `README` with this example: `STAN_DOCS_VERSION=2.34 quarto render src/functions-reference/ --output functions_
reference.pdf`

mitzimorris replied:
Added `subtitle` to _quarto.yml's  `format` section.    (you'd think this would go with `title` in the `project` section, but you'd be wrong.  wtf???)

mitzimorris replied:
I think that there are enough fiddly details that a script might help.

WardBrian commented on src/_quarto.yml:
```diff
@@ -0,0 +1,255 @@
+toc-depth: 3
+
+website:
+  favicon: img/logo_tm.svg
```
I get a warning that this svg doesn't exist

mitzimorris replied:
fixed

WardBrian commented on src/functions-reference/index.qmd:
Not sure if it should belong here or somewhere else, but I noticed the PDFs built from this branch don't have the "Version 2.34" subheader or anything to indiciate which version they are (e.g., they ignore `STAN_DOCS_VERSION`)

Is this possible in quarto? Even if we can't have it as a pdf subtitle, it would be nice to list somewhere in this page still

mitzimorris replied:
see above comment

WardBrian commented on src/stan-users-guide/quarto_styles.css:
Why must there be several `quarto_styles.css` files? Are they different from each other?

mitzimorris replied:
good catch - not needed.

WardBrian reviewed (changes requested):
Thanks for plugging away on this. A few questions/concerns about the current state of the PR

mitzimorris commented on src/stan-users-guide/part3.Rmd:
```diff
@@ -1,4 +1,4 @@
-# <i style="font-size: 110%; color:#990017;">Part 3. Posterior Inference & Model Checking</i> {- #meta-models.part}
+# <i style="font-size: 110%; color:#052744;">Part 3. Posterior Inference & Model Checking</i> {.unnumbered #meta-models.part}
```
defined here:  https://github.com/stan-dev/docs/blob/299e90e13240932edaaf4c8f034fbf6247f0fd55/src/quarto_styles.css#L3-L12.  (redundancies because I was trying different things)

WardBrian replied:
In that case I think you can use `color: var(--bs-body-color)` instead of the hex code, which would make any changes in the future easier to sync?

mitzimorris commented on src/stan-users-guide/clustering.Rmd:
```diff
@@ -366,7 +369,7 @@ model {
   }
   for (n in 1:N) {
-    for (k in 1:K) {
+    for (k in 1:K) { \\
```
good catch - thanks!

mitzimorris commented on src/functions-reference/simplex_distributions.Rmd:
```diff
@@ -8,23 +8,19 @@ specified $K$.  A $K$-dimensional vector $\theta$ is a unit
+\end{equation*}
 
-_**Warning:**_  If any of the components of $\theta$ satisfies
+\emph{Warning:}  If any of the components of $\theta$ satisfies
```
changed back - works everywhere.  not sure why it was changed.

mitzimorris commented on src/js.html:
comes from here:
https://github.com/quarto-dev/quarto-web/blob/main/js.html

don't know if we need it - it was imported along with the rest of the quarto website styling/files.

WardBrian commented on src/_quarto.yml:
```diff
@@ -0,0 +1,255 @@
+    - "reference-manual/*.Rmd"
+    - "stan-users-guide/*.Rmd"
+    - "!sdlc.Rmd"
+    - "!cmdstan-guide/index.Rmd"
+    - "!functions-reference/index.Rmd"
+    - "!reference-manual/index.Rmd"
+    - "!stan-users-guide/index.Rmd"
```
Is there a reason we now have two `index` files?

mitzimorris replied:
only if we want to be able to build with bookdown - the yaml header in the index.Rmd contains a bunch of formatting info.

WardBrian replied:
I guess for an initial merge it is fine to leave it. Once we sort out the kinks I see no reason to continue to support bookdown after that

WardBrian reviewed (changes requested):
A few more comments, but mostly more around config and wondering if we can do some things in a nicer way.

I don't know if I just now noticed this or if it wasn't like this before, but when I build the web version with the latest changes the top bar links to each doc set are drop downs, which don't really work that well with the number of sub-pages we have. Personally I'd prefer if all of them except "Interfaces" were just straight links to the overview page. If you want a page that isn't in the first ~15, you have to go there anyway, since you can't scroll the dropdowns.

mitzimorris commented on src/404.Rmd:
OK, will revert back.

mitzimorris commented:
big problem - quarto website vs individual pdfs per document directory - I'm not sure we can get links that work for both.

WardBrian commented:
What do you mean?

mitzimorris commented:
> What do you mean?

updated previous comment.   bottom line:   this is tedious, but doable.  best done when brain dead.

mitzimorris commented:
> I don't know if I just now noticed this or if it wasn't like this before, but when I build the web version with the latest changes the top bar links to each doc set are drop downs, which don't really work that well with the number of sub-pages we have.

agree. 

at some point I changed the config so that the nav bar entries mirrored the corresponding sidebar entries.   bad idea.

mitzimorris commented:
@WardBrian - ready for another review pass.

mitzimorris commented on src/functions-reference/_quarto.yml:
```diff
@@ -0,0 +1,74 @@
+    pdf-engine: "pdflatex"
+    latex-engine: "pdflatex"
+    documentclass: book
+    subtitle: "Version {{< env STAN_DOCS_VERSION >}}"
```
done!

WardBrian reviewed (changes requested):
A bunch of individual comments. The new links all look good, thanks. 

There are also the following hard-links (not relative) that will break due to the output files having new names. Should be a simple find and replace for the most part:

```
mc-stan.org/docs/functions-reference/functions-map.html
mc-stan.org/docs/stan-users-guide/stanc-pretty-printing.html
mc-stan.org/docs/functions-reference/deprecated-functions.html
mc-stan.org/docs/reference-manual/removed-features.html
mc-stan.org/docs/stan-users-guide/stand-alone-generated-quantities-and-ongoing-prediction.html
mc-stan.org/docs/stan-users-guide/index.html
mc-stan.org/docs/reference-manual/function-names.html
mc-stan.org/docs/stan-users-guide/understanding-stanc3-errors-and-warnings.html
mc-stan.org/docs/functions-reference/index.html
mc-stan.org/docs/reference-manual/language-syntax.html
mc-stan.org/docs/reference-manual/notation-for-samples-chains-and-draws.html
mc-stan.org/docs/reference-manual/data-types.html
mc-stan.org/docs/stan-users-guide/posterior-predictive-simulation-in-stan.html
mc-stan.org/docs/stan-users-guide/model-conditioning-and-curvature.html
mc-stan.org/docs/reference-manual/variable-transforms.html
mc-stan.org/docs/reference-manual/variable-declaration.html
mc-stan.org/docs/reference-manual/deprecated-features.html
mc-stan.org/docs/reference-manual/stochastic-gradient-ascent.html
mc-stan.org/docs/stan-users-guide/stanc-args.html
mc-stan.org/docs/reference-manual/program-block-generated-quantities.html
mc-stan.org/docs/reference-manual/extra-grammatical-constraints.html
mc-stan.org/docs/functions-reference/functions-1d-integrator.html
mc-stan.org/docs/reference-manual/bfgs-and-l-bfgs-configuration.html
mc-stan.org/docs/stan-users-guide/using-the-stan-compiler.html
mc-stan.org/docs/reference-manual/index.html
mc-stan.org/docs/reference-manual/language-multi-indexing.html
mc-stan.org/docs/reference-manual/hmc.html
mc-stan.org/docs/reference-manual/hmc-algorithm-parameters.html
mc-stan.org/docs/reference-manual/effective-sample-size.html
mc-stan.org/docs/reference-manual/licensing.html
```

mitzimorris commented on src/_quarto.yml:
```diff
@@ -0,0 +1,263 @@
+            url: https://mc-stan.org/rstan
+          - text: "RStanArm"
+            url: https://mc-stan.org/rstanarm
+          - text: "Stan.jl"
```
sure.  list is in alphabetical order - I think that puts BridgeStan fisrt!

WardBrian replied:
Might be worth splitting into "Stan interfaces"  (Rstan, etc) and "Related packages" (BridgeStan, BRMS, loo?, bayesplot?)

mitzimorris commented on src/_quarto.yml:
```diff
@@ -70,12 +74,14 @@ website:
     - id: stan-users-guide 
-      title: "User's Guide"
+      title: "Stan User's Guide"
+      subtitle: "Hi Mom!"
```
gone - looks like there's a bug in quarto such that this didn't ever render.

mitzimorris commented on src/_quarto.yml:
```diff
@@ -0,0 +1,263 @@
+            url: https://mc-stan.org/rstanarm
+          - text: "Stan.jl"
+            url: http://stanjulia.github.io/Stan.jl/stable/INTRO/
+    tools:
+      - icon: github
+        url: https://github.com/stan-dev
+      - icon: mastodon
+        href: "https://bayes.club/%40mcmc_stan"
+      - icon: youtube
+        href: "https://www.youtube.com/channel/UCwgN5srGpBH4M-Zc2cAluOA"
```
sure.  trying to figure out the icon bit.

mitzimorris replied:
for the record - hit this bug:  https://github.com/quarto-dev/quarto-cli/issues/6359

there is a font-awesome free svg for discourse, but the website handling of the navbar tools is too brittle to be able to use it.

using the supported bootstrap icon for "chat".

mitzimorris replied:
another bug - no way to link to mastodon - https://github.com/quarto-dev/quarto-cli/issues/3493

encoding `@` as "%40" (unicode code point) breaks mastodon
leaving `@` in somehow gets processed as a (missing) citeproc - resulting URL contains "{?mcmc-stan}".

mitzimorris commented on src/functions-reference/binary_distributions.Rmd:
```diff
@@ -76,7 +76,7 @@ function of y given chance of success `theta`
 used in transformed data and generated quantities blocks.
 For a description of argument and return types, see section
-[vectorized PRNG functions](#prng-vectorization).
+[vectorized PRNG functions](conventions_for_probability_functions.Rmd#prng-vectorization).
```
I see no reason to use bookdown after this PR.

bob-carpenter replied:
> I see no reason to use bookdown after this PR.

Yay!

WardBrian replied:
In that case we can [avoid some duplication](https://github.com/stan-dev/docs/pull/694#discussion_r1481806339)

mitzimorris commented on src/stan-users-guide/user-functions.Rmd:
if you have wider browser tabs, you'll see that you have the website project title ("Stan Documentation"), followed by the internal anchor name of the chapter or section.

investigating options.

mitzimorris replied:
added "pagetitle" to yaml for each page.
in CmdStan Guide, tweaked pagetitle to get most salient info up front

WardBrian commented:
The latest commit has fixed the tab titles for me for the SUG, but the other 3 manuals still don't have the page title in the tab

WardBrian commented:
We shouldn't use relative links between pdfs (it doesn't currently for bookdown and I see no way this would ever work sensibly). In the current docs we just use full links to mc-stan.org/docs/.... when we need to link between two different manuals

mitzimorris commented:
OK, full links between manuals everywhere.

mitzimorris commented:
ready for review.   then we need to figure out how to launch the monolithic site.

mitzimorris commented:
@WardBrian - this time for sure?   double checked links, got rid of use of html anchor ids.

WardBrian commented on src/cmdstan-guide/overview.qmd:
```diff
@@ -0,0 +1,19 @@
+
+::: {.g-col-11}
+
+{{< include preface.qmd >}}
```
This is great!

WardBrian commented on src/cmdstan-guide/print.Rmd:
```diff
@@ -1,3 +1,7 @@
+---
+pagetitle: Deprecations
+---
+
```
Maybe the wrong title?

mitzimorris replied:
fixed - I think that there might have been a `Deprecated` section that got lost.

WardBrian commented on src/stan-users-guide/simulation-based-calibration.Rmd:
````diff
@@ -410,21 +414,21 @@ After running this for enough iterations so that the effective sample
 the ranks are computed and binned, and then plotted.
 
-```{r include = TRUE, echo = FALSE, fig.align = "center", fig.cap = "Simulation based calibration plots for location and scale  of a normal model with standard normal prior on the location  standard lognormal prior on the scale.  Both histograms appear uniform, which is consistent with inference being well calibrated."}
+```{r include = TRUE, echo = FALSE, out.width="90%", fig.align = "center", fig.cap = "Simulation based calibration plots for location and scale of a normal model with standard normal prior on the location, standard lognormal prior on the scale.  Both histograms appear uniform, which is consistent with inference being well calibrated."}
````
Later in this file you changed the figures to not use knitr, thoughts on doing it for the rest?

mitzimorris replied:
eliminated.

WardBrian reviewed (commented):
Looking really good! Sorry it took a while this round, a lot of files to check

WardBrian commented on src/_quarto.yml:
```diff
@@ -0,0 +1,272 @@
+
+bibliography: bibtex/all.bib
+biblio-style: apalike
+syntax-definition: "stan.xml"
```
`highlight-theme: tango` in the `html` section seems to get back our current theme, which still looks good IMO. Probably something similar in pdf?

mitzimorris replied:
good catch - looking at quarto-generated website and pdfs, it's not clear
a) whether or not `stan.xml` is registered with pandoc
b) what the correct directive for syntax highlighting is.
investigating.

WardBrian replied:
I know pandoc supports Stan. Last I checked the only difference with our custom Stan.xml is just the tuple keyword. 

It looks like the highlighting is happening, the default colors are just washed out. The tango theme seems better to me

mitzimorris replied:
quarto config was wrong -   in the `format` section,  need to have both `syntax-definitions` and `highlight-style` options specified.    with these changes, syntax coloring in both pdfs and html pages matches current website syntax highlighting.

mitzimorris replied:
pushed changes to configs.

mitzimorris commented:
should I add cleanup to this PR by getting rid bookdown config files `_output.yml` `index.Rmd`, `book.json` ?

WardBrian commented:
You could, we will probably need some follow up PRs anyway at least for Jenkins, the readme, etc, so I think it’s also reasonable to wait

WardBrian commented:
@mitzimorris the reference manual's quarto.yml got missed for the highlighting updates

WardBrian commented:
@mitzimorris are you going to delete old `index.Rmd`s here or would you rather leave that for another PR? Going to try to review again tomorrow morning

mitzimorris commented:
I can do a cleanup commit.

mitzimorris commented:
cleaned out all files not needed to use quarto to build the docs and modified the quarto build script to build the website.
just realized the quarto build script args need to be reworked because the logic is off; the simplest fix would be either build the unified html website, or build one or more pdfs.   will try to fix tonight.

mitzimorris commented:
another thing that needs updating - links to latest.

WardBrian commented:
I think it sufficies to decide where we want to put the link on the new pages and update the script

for the existing pages, it will be a similar process to what I just did in #709. For various reasons it may be easier to wait to do those until the new version is actually live online

mitzimorris commented on src/reference-manual/language.Rmd:
yes, it's now in the preface.

mitzimorris commented on build_quarto.py:
```diff
@@ -0,0 +1,119 @@
+    if (len(sys.argv) > 4):
+        print("Unused arguments:  %s" % ' '.join(sys.argv[4: ]))
+
+    if (build_web):
```
rename makes sense.   as part of cleanup?

WardBrian replied:
Since the original `build.py` is now un-runnable, I don't think it should survive the merge of this

WardBrian reviewed (commented):
A few problems with build.py

I also cannot get the users-guide to build as a pdf - just me? The log file is full of broken references (which we should fix, e.g. `LaTeX Warning: Reference `fig-theta1' on page 360 undefined on input line 18388.`) but no clear error

Edit: Seemed like it was some weird whitespacing error with one of the equations. 🤷🏻

WardBrian commented:
I just pushed a change to `link_to_latest.py`, so I think that the only change needed for jenkins is to add Quarto to the docker image at https://github.com/stan-dev/ci-scripts/blob/master/docker/docs/Dockerfile

@serban-nicusor-toptal would you mind doing that? We can get a deb from https://quarto.org/docs/get-started/


So, after the above changes to build.py, I think we can:

1. Merge this
2. Re-build the 2.35 docs, based on Quarto. 
3. Once those are built, we can do another pass using the automated link checker tools to update any places that need to point to the new urls Quarto will give them

WardBrian commented:
Oh, and we need to rewrite the readme.

serban-nicusor-toptal commented:
> I just pushed a change to `link_to_latest.py`, so I think that the only change needed for jenkins is to add Quarto to the docker image at https://github.com/stan-dev/ci-scripts/blob/master/docker/docs/Dockerfile
> 
> @serban-nicusor-toptal would you mind doing that? We can get a deb from https://quarto.org/docs/get-started/
Tested and pushed to master, see https://github.com/stan-dev/ci-scripts/commit/36f3184662c41fb6a3a3524ecdde91b1693f4bfa

mitzimorris commented:
users guide builds for me?


```
(base) ~/github/stan-dev/docs (issue/689-quarto-pdfs)> pdflatex --version
pdfTeX 3.141592653-2.6-1.40.25 (TeX Live 2023)
kpathsea version 6.3.5
...
Compiled with libpng 1.6.39; using libpng 1.6.39
Compiled with zlib 1.2.13; using zlib 1.2.13
Compiled with xpdf version 4.04
(base) ~/github/stan-dev/docs (issue/689-quarto-pdfs)> quarto --version
1.3.450
```

WardBrian commented:
Have you already pulled? I fixed whatever issue I was having in [4c6d8c9](https://github.com/stan-dev/docs/pull/694/commits/4c6d8c99059bcd7e8af343bf2fc10c3e35bc5c11)

mitzimorris commented on build.py:
```diff
@@ -28,101 +38,98 @@ def pushd(new_dir):
         sys.exit(1)
 
-    stan_version = '_'.join([str(stan_major), str(stan_minor)])
+    stan_version = "_".join([str(stan_major), str(stan_minor)])
```
with the script, no longer need file `src/_environment`

WardBrian replied:
Not sure if that file worked in the PDFs at any rate

mitzimorris commented:
pulled, merged - next up, README

WardBrian reviewed (approved):
Thanks for all the hard work!

Let's see how this goes
